paasta_tools/deployd/watchers.py

```diff
@@ -45,6 +55,130 @@ def startup_checker(self, obj):
         self.is_ready = True
 
 
+class PublicConfigFileWatcher(PaastaWatcher):
+
+    def __init__(self, inbox_q, cluster, **kwargs):
+        super(PublicConfigFileWatcher, self).__init__(inbox_q, cluster)
+        self.wm = pyinotify.WatchManager()
+        self.wm.add_watch(PATH_TO_SYSTEM_PAASTA_CONFIG_DIR, self.mask, rec=True)
+        self.notifier = pyinotify.Notifier(watch_manager=self.wm,
+                                           default_proc_fun=PublicConfigEventHandler(filewatcher=self))
+
+    @property
+    def mask(self):
+        boring_flags = ['IN_CLOSE_NOWRITE', 'IN_OPEN', 'IN_ACCESS', 'IN_ATTRIB']
+        return reduce(lambda x, y: x | y,
+                      [v for k, v in pyinotify.EventsCodes.OP_FLAGS.items() if k not in boring_flags])
+
+    def run(self):
+        self.notifier.loop(callback=self.startup_checker)
+
+    def startup_checker(self, obj):
+        self.is_ready = True
+
+
+def get_marathon_apps():
+    marathon_config = load_marathon_config()
+    marathon_client = get_marathon_client(marathon_config.get_url(), marathon_config.get_username(),
+                                          marathon_config.get_password())
+    return get_all_marathon_apps(marathon_client, embed_failures=True)
+
+
+class MaintenanceWatcher(PaastaWatcher):
+    def __init__(self, inbox_q, cluster, **kwargs):
+        super(MaintenanceWatcher, self).__init__(inbox_q, cluster)
+        self.draining = []
+
+    def run(self):
+        self.is_ready = True
+        while True:
+            draining_hosts = get_draining_hosts()
+            new_draining_hosts = [host for host in draining_hosts if host not in self.draining]
+            self.draining = draining_hosts
+            service_instances = []
+            if new_draining_hosts:
+                self.log.info("Found new draining hosts: {}".format(new_draining_hosts))
+                service_instances = self.get_at_risk_service_instances(new_draining_hosts)
+            for service_instance in service_instances:
+                self.inbox_q.put(service_instance)
+            time.sleep(20)
+
+    def get_at_risk_service_instances(self, draining_hosts):
+        marathon_apps = get_marathon_apps()
+        at_risk_tasks = [task for app in marathon_apps for task in app.tasks if task.host in draining_hosts]
+        self.log.info("At risk tasks: {}".format(at_risk_tasks))
+        service_instances = []
+        for task in at_risk_tasks:
+            app_id = task.app_id.strip('/')
+            service, instance, _, __ = deformat_job_id(app_id)
+            service_instances.append(ServiceInstance(service=service,
```

Comment:
That's a good point. There is some de-duping in the `Inbox` logic which probably means we are okay...

It might be sane to de-dupe here though anyway. I think it would be clearer that way...